### ISSUE TITLE
NaN loss with zero gradients (in first iter) for PSGD since 1.7.0

### ISSUE THREAD

**Issue by JCBrouwer** [HomebrewML/HeavyBall#55]
I don't have an easy to extract reproduction case, but since March 26th's update to 1.7.0 training with PSGD leads to NaN parameters after the first .step().

I printed the gradients of parameters just before .step() and it seems they are close to / exactly zero. Maybe the new update has introduced some subtle divide by zero issue?

Reverting back to 1.6.3 fixes the issue for me with no further differences.

**Comment by ClashLuke** [comment]
Thank you for the report! I'm looking into it now. I could reproduce it with Newton's method + finite differences, but not anything else. Is that your setup as well?

**Comment by JCBrouwer** [comment]
I'm using `ForeachPSGDKron` with learning rate, weight decay, and warmup steps specified.

I'm also using `heavyball.utils.compile_mode = "default"`

Other than that I haven't changed any settings or defaults.

Does that use newton's method / finite differences under the hood?

**Comment by ClashLuke** [comment]
Oh, interesting! You can recover the previous convergence by setting `precond_init_scale=1`. 

1.7.0 introduced a method of setting the `precond_init_scale`, which should strictly improve convergence. However, it seems like it didn't for you.

It's set as `1 / mean(grad**4) ** 8`, so large gradients and zero-gradients will cause a NaN. Are you facing either of these?

---

Regarding Newton's method: it's only used for `Newton*` optimizers like `ForeachCachedNewtonPSGD`. Those would compute the hessian vector product (via finite differences or using two backward passes) for a more accurate optimizer step at increased cost and decreased flexibility.

**Comment by ClashLuke** [comment]
Please try upgrading to 1.7.2 and let me know if it's resolved.

I've implemented a new, significantly more stable method to compute precond_init_scale that might've fixed your issues. If not, it'll fail loudly and give direct, applicable advice.

The first PSGD step now has an added overhead (due to fp64 computation and .item() calls for sanity checks and error reporting), but the restructured PSGD implementation should still speed things up (I'm seeing 10% _runtime_ advantage after 300k steps in toy problems).

**Comment by ClashLuke** [comment]
Closing due to inactivity. Please reopen the issue if it's not fixed!
